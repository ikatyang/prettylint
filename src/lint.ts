import * as path from 'path';
import * as prettier from 'prettier';
import { diagnose } from './diagnose';

// https://eslint.org/docs/developer-guide/working-with-custom-formatters#description-of-the-results
export interface LintResult {
  /**
   * The path to the file relative to the current working directory (the path from which eslint was executed).
   */
  filePath: string;
  /**
   * An array of message objects. See below for more info about messages.
   */
  messages: LintMessage[];
  /**
   * The number of errors for the given file.
   */
  errorCount: number;
  /**
   * The number of warnings for the given file.
   */
  warningCount: number;
  /**
   * The source code for the given file. This property is omitted if this file has no errors/warnings or if the `output` property is present.
   */
  source: string;
  /**
   * The source code for the given file with as many fixes applied as possible. This property is omitted if no fix is available.
   */
  output?: string;
}

export const enum LintSeverity {
  Warning = 1,
  Error = 2,
}

export interface LintMessage {
  /**
   * the id of the rule that produced the error or warning.
   */
  ruleId: string;
  /**
   * the severity of the failure, 1 for warnings and 2 for errors.
   */
  severity: LintSeverity;
  /**
   * the human readable description of the error.
   */
  message: string;
  /**
   * the line where the issue is located.
   */
  line: number;
  /**
   * the column where the issue is located.
   */
  column: number;
  /**
   * the type of the node in the AST
   */
  nodeType: string;
}

export interface LintOptions {
  config?: boolean | string;
}

export async function lint(
  filename: string,
  content: string,
  options: LintOptions = {},
): Promise<LintResult> {
  const { config = true } = options;

  const absolute_filename = path.resolve(filename);
  const relative_filename = path.relative('.', filename);

  const node_type = 'Unknown';
  const rule_id = 'prettier/prettier';

  try {
    const prettier_options =
      config === false
        ? null
        : await prettier.resolveConfig(absolute_filename, {
            config: config === true ? undefined : config,
          });

    const { diagnostics, formatted } = diagnose(content, {
      filepath: filename,
      ...prettier_options,
    });

    const messages = diagnostics.map(
      ({ message, start: { line, column } }): LintMessage => ({
        ruleId: rule_id,
        severity: LintSeverity.Warning,
        line,
        column,
        message,
        nodeType: node_type,
      }),
    );

    return {
      filePath: relative_filename,
      source: content,
      ...messages.length === 0 ? {} : { output: formatted },
      messages,
      errorCount: 0,
      warningCount: messages.length,
    };
  } catch (error) {
    const message: LintMessage = {
      ruleId: rule_id,
      severity: LintSeverity.Error,
      nodeType: node_type,
      ...error.loc
        ? // SyntaxError from parser in prettier
          {
            line: error.loc.start.line,
            column: error.loc.start.column,
            message: error.message.split('\n', 1)[0], // cut the code-frame
          }
        : // istanbul ignore next
          {
            line: 0,
            column: 0,
            message:
              /* istanbul ignore next */ typeof error.message === 'string'
                ? error.message
                : error,
          },
    };
    return {
      filePath: relative_filename,
      source: content,
      messages: [message],
      errorCount: 1,
      warningCount: 0,
    };
  }
}
