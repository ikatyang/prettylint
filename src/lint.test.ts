import * as path from "node:path";
import { expect, test } from "vitest";
import { lint } from "./lint.js";

const baseDir = path.resolve(__dirname, "../fixtures/lint");

test("filename: any, config: false", async () => {
  const result = await lint("path/to/somewhere.js", "const n = 0;", {
    config: false,
  });
  expect(result.output).toBe("const n = 0;\n");
});

test('filename: any, config: "path/to/singleQuote/.prettierrc"', async () => {
  const config = `${baseDir}/singleQuote/.prettierrc`;
  const result = await lint("path/to/somewhere.js", 'let x = "hi";', {
    config,
  });
  expect(result.output).toBe("let x = 'hi';\n");
});

test('filename: "path/to/useTabs/file.js", config: undefined', async () => {
  const filename = `${baseDir}/useTabs/file.js`;
  const result = await lint(filename, "if (true) { console.log(123); }");
  expect(result.output).toBe("if (true) {\n\tconsole.log(123);\n}\n");
});

test("content: SyntaxError", async () => {
  const result = await lint("path/to/somewhere.js", "const 123;", {
    config: false,
  });
  expect(result).toMatchSnapshot();
});

test("content: formatted", async () => {
  const result = await lint("path/to/somewhere.js", "let x = 1;\n", {
    config: false,
  });
  expect(result).toMatchSnapshot();
});

test('filename: "path/to/filepath/example.json"', async () => {
  const result = await lint("path/to/somewhere.json", '{\n"name": "hello"\n}');
  expect(result.output).toBe('{\n  "name": "hello"\n}\n');
});
