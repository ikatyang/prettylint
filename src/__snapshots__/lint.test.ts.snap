// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`failure 1`] = `
{
  "errorCount": 1,
  "filePath": "path/to/somewhere.js",
  "messages": [
    {
      "column": 7,
      "line": 1,
      "message": "Unexpected token (1:7)",
      "nodeType": "Unknown",
      "ruleId": "prettier/prettier",
      "severity": 2,
    },
  ],
  "source": "const 123;",
  "warningCount": 0,
}
`;

exports[`success 1`] = `
{
  "errorCount": 0,
  "filePath": "path/to/somewhere.js",
  "messages": [],
  "source": "let x = 1;
",
  "warningCount": 0,
}
`;
